src/monodroid/jni/strings.hh

```diff
@@ -0,0 +1,731 @@
+			append_integer (buffer.get (), i);
+			return *this;
+		}
+
+		string_base& append (uint16_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (i);
+			return *this;
+		}
+
+		string_base& append (int32_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (i);
+			return *this;
+		}
+
+		string_base& append (uint32_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (i);
+			return *this;
+		}
+
+		string_base& append (int64_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (i);
+			return *this;
+		}
+
+		string_base& append (uint64_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (i);
+			return *this;
+		}
+
+		force_inline string_base& assign (const TChar* s, size_t length) noexcept
+		{
+			idx = 0;
+			return append (s, length);
+		}
+
+		string_base& assign (const char *s) noexcept
+		{
+			if (s == nullptr)
+				return *this;
+
+			return assign (s, strlen (s));
+		}
+
+		force_inline bool next_token (size_t start_index, const TChar separator, string_segment& token) const noexcept
+		{
+			size_t index;
+			if (token._fresh) {
+				token._fresh = false;
+				token._last_index = start_index;
+				index = start_index;
+			} else {
+				index = token._last_index + 1;
+			}
+
+			token._start = nullptr;
+			token._length = 0;
+			if (token._last_index + 1 >= buffer.size ()) {
+				return false;
+			}
+
+			TChar *start = buffer.get () + index;
+			TChar *p = start;
+			while (*p != NUL) {
+				if (*p == separator) {
+					break;
+				}
+				p++;
+				index++;
+			}
+
+			token._last_index = *p == NUL ? buffer.size () : index;
+			token._start = start;
+			token._length = static_cast<size_t>(p - start);
+
+			return true;
+		}
+
+		force_inline bool next_token (const char separator, string_segment& token) const noexcept
+		{
+			return next_token (0, separator, token);
+		}
+
+		force_inline ssize_t index_of (const TChar ch) const noexcept
+		{
+			TChar *p = buffer.get ();
+			while (p != nullptr && *p != NUL) {
+				if (*p == ch) {
+					return static_cast<ssize_t>(p - buffer.get ());
+				}
+				p++;
+			}
+
+			return -1;
+		}
+
+		force_inline bool starts_with (const TChar *s, size_t s_length) noexcept
+		{
+			if (s == nullptr || s_length == 0 || s_length > buffer.size ())
+				return false;
+
+			return memcmp (buffer.get (), s, s_length) == 0;
+		}
+
+		force_inline bool starts_with (const char* s) const noexcept
+		{
+			if (s == nullptr)
+				return false;
+
+			return starts_with (s, strlen (s));
+		}
+
+		force_inline void set_length_after_direct_write (size_t new_length) noexcept
+		{
+			set_index (new_length);
+			terminate ();
+		}
+
+		force_inline void set_at (size_t index, const TChar ch) noexcept
+		{
+			if (index >= buffer.size ())
+				return;
+
+			TChar *p = buffer + index;
```

Comment:
I haven't traced through the semantics, but can't `this->idx` differ from `buffer.size()`?  Wouldn't it be possible for `this->idx` to be *before* `buffer+index`, meaning we're attempting to write *after* the current `NUL`?  Is that desirable?

Sure, this class has a comment saying "it does no bounds checking; child classes should do that", but *this* method is (1) not `virtual`, and (2) `public`, meaning it is directly usable regardless of class.

Rephrased: does allowing this make any sense?

```cpp
static_local_string<10> s;
for (int i = 0; i < 10; ++i) s.set_at (i, '0' + i);
s.set_index(0);
s.set_at (5, 'a');
```

Reply:
It makes sense, hence the presence of `length()` and `index()` methods you mentioned earlier. However, the question is - do we *need* the functionality in your code excerpt? My initial idea here was to implement only what we actually use and add functionality when need arises